### PR TITLE
step_1_2 

### PR DESCRIPTION
**Pull request by olgZZZ**
---
name: Step_1_2
about: Box and Pin
---

### CODE REVIEW

File: 1_concepts/1_2_box_pin/src/main.rs [hunk 1]
```diff
@@ -1,4 +1,145 @@
...
+impl MutMeSomehow for String 
+{
+  fn mut_me_somehow ( mut self : Pin< &mut Self > )
+  {
+    self.set( "new_str".to_string() );
+  }
+}
+
+impl SayHi for String {}
+
```
Comment (dmvict) [inline 1_concepts/1_2_box_pin/src/main.rs:65]: The traits is not implemented for several types from the list 
> Box\<T\>, Rc\<T\>, Vec\<T\>, String, &[u8], T

Please, implement it.

File: 1_concepts/1_2_box_pin/src/main.rs [hunk 2]
```diff
@@ -1,4 +1,145 @@
...
+    let val = Rc::new( "str" );
+    Pin::new( &val ).say_hi();
+  }
+
+  // tests for Vec< T >
+
+  #[ test ]
+  fn for_vec_mutme()
+  {
+    let mut val = vec![1; 3];
```
Comment (dmvict) [inline 1_concepts/1_2_box_pin/src/main.rs:116]: Please, update formatting.

File: 1_concepts/1_2_box_pin/src/main.rs (hunk 1 above)
Comment (olgZZZ) [reply 1_concepts/1_2_box_pin/src/main.rs:65]: Не получается для типа  - Т реализовать. Конфликт имплементаций с Box< T >

File: 1_concepts/1_2_box_pin/src/main.rs [hunk 3]
```diff
@@ -1,4 +1,222 @@
...
+    Pin::new( &val ).say_hi();
+  }
+
+  // tests for Rc< T >
+
+  #[ test ]
+  fn for_rc_mutme()
+  {
+    let t = "str";
+    let mut val = Box::new( t );
```
Comment (dmvict) [inline 1_concepts/1_2_box_pin/src/main.rs:142]: Please, write test for Rc.

**Comment by dmvict** [comment]
Can be merged.

**Comment by dmvict** [comment]
Can be merged.

**Pull request merged**